locust/dispatch.py

```diff
@@ -49,548 +48,134 @@ class UsersDispatcher(Iterator):
-                if self._users_left_to_assigned[worker_node_id][user_class_to_add] == 0:
-                    continue
-                if self._try_next_worker_in_order_to_stay_balanced_during_ramp_up(worker_node_id, user_class_to_add):
-                    continue
-                self._dispatched_users[worker_node_id][user_class_to_add] += 1
-                self._users_left_to_assigned[worker_node_id][user_class_to_add] -= 1
-                user_count_in_current_dispatch += 1
-                break
-
-            if user_count_in_current_dispatch == self._user_count_per_dispatch:
-                break
-
-            if self._dispatched_user_count == self._desired_user_count:
-                break
-
-        return {
-            worker_node_id: dict(sorted(user_classes_count.items(), key=itemgetter(0)))
-            for worker_node_id, user_classes_count in sorted(self._dispatched_users.items(), key=itemgetter(0))
-        }
-
-    def _crash_dump(self) -> Tuple[str, str]:
-        """Parameters necessary to debug infinite loop issues.
-
-        Users encountering an infinite loop issue should provide these informations.
-        """
-        crash_dump = json.dumps(
-            {
-                "spawn_rate": self._spawn_rate,
-                "initial_dispatched_users": self._initial_dispatched_users,
-                "desired_users_assigned_to_workers": self._desired_users_assigned_to_workers,
-                "user_classes_count": self._user_classes_count,
-                "initial_user_count": sum(map(sum, map(dict.values, self._initial_dispatched_users.values()))),
-                "desired_user_count": sum(self._user_classes_count.values()),
-                "number_of_workers": self._number_of_workers,
-            },
-            indent="  ",
-        )
-        fp = tempfile.NamedTemporaryFile(
-            prefix="locust-dispatcher-crash-dump-", suffix=".json", mode="wt", delete=False
-        )
-        try:
-            fp.write(crash_dump)
-        finally:
-            fp.close()
-        return crash_dump, fp.name
-
-    @property
-    def _number_of_workers(self) -> int:
-        return len(self._users_left_to_assigned)
+    def __next__(self) -> Dict[str, Dict[str, int]]:
+        return next(self._dispatcher_generator)
 
-    @property
-    def _all_users_have_been_dispatched(self) -> bool:
-        return self._user_count_left_to_assigned == 0
+    def _dispatcher(self) -> Generator[Dict[str, Dict[str, int]], None, None]:
+        # TODO: Ensure initial users on workers are ok. If not, send a single dispatch
+        #       to set the workers with the adequate users. After having done that,
+        #       we need to update `self._users_on_workers` and `self._current_user_count`.
 
-    @property
-    def _user_count_left_to_assigned(self) -> int:
-        return sum(map(sum, map(dict.values, self._users_left_to_assigned.values())))
+        if self._current_user_count < self._target_user_count:
```

Comment:
this if and elif statements seem unnecessary?

Reply:
```python
        if self._current_user_count == self._target_user_count:
            yield self._initial_users_on_workers
            return

        while self._current_user_count < self._target_user_count:
            with self._wait_between_dispatch_iteration_context():
                yield self._ramp_up()

        while self._current_user_count > self._target_user_count:
            with self._wait_between_dispatch_iteration_context():
                yield self._ramp_down()
```

better?